doc/design/inference.md

```diff
@@ -0,0 +1,178 @@
+
+## Inference Program
+
+A simple inference program may be defined in Python API as:
+
+```python
+image = fluid.layers.data(name='x', shape=[784], dtype='float32')
+predict = fluid.layers.fc(input=image,
+                          size=10,
+                          act='softmax')
+```
+
+After training for several epochs, the parameters can be saved using the method [fluid.io.save_inference_model](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/fluid/io.py), which will save the binary proto string of the program at the same time.
+
+```python
+fluid.io.save_inference_model(
+                "./inference_model/", ["x"], [predict],
+                exe)
+```
+
+Like training, there is a `main_program` and a `startup_program` for inference.
+- The `main_program` defines the computational operators and all variables and can be evaluated as many times as users want. The protobuf message of it is saved by `fluid.io.save_inference_model`. Thus, it can be initilized from file or a pre-loaded buffer.
+- The `startup_program` program is responsible for initializing all parameters. Since all the parameters are saved to files, the `startup_program` is composed of `load_op`s and needs to be evaluated for a specified executor only one time. There is no need to save the protobuf message of the `startup_program` because it can be easily derived from the `main_program`. The `startup_program`.
+
+### Introduce of ProgramBuilder
+
+We introduce the concept of `ProgramBuilder`, which will collect all the metadata of inference program and support transform and optimization of the inference program.
+
+```cpp
+class ProgramBuilder {
+ public:
+  // Initialize from file
+  ProgramBuilder(const std::string& filename);
+  // Initialize from buffer
+  ProgramBuilder(const char* buffer, const size_t num_bytes);
+
+  // Some utility interface maybe required by users
+  std::vector<std::string>& GetFeedVarNames() const;
+  std::vector<std::string>& GetFecthVarNames() const;
+  std::vector<int64_t> GetFeedVarShape(const size_t index);
+  std::vector<int64_t> GetFetchVarShape(const size_t index);
+
+  void AppendFetchVariables(const std::string& var_name);
+  ...
+
+  // Do transform to the inference program
+  ProgramBuilder* operator()(/* some optimizing strategy */);
+  ProgramBuilder* operator()(const std::vector<std::string>& feed_var_names,
+                             const std::vector<std::string>& fetch_var_names,
+                             /* some optimizing strategy */);
+
+ private:
+  framework::ProgramDesc* main_program_;
+  framework::ProgramDesc* startup_program_;
+  std::vector<std::string> feed_var_names_;
+  std::vector<std::string> fetch_var_names_;
+};
+```
+
+In the first design, `ProgramBuilder` contains all the elements memtioned above, and is instanced by protobuf message of the `main_program`. Other members `startup_program`, `feed_var_names` and `fetch_var_names` will also be derived in the constructor.
```

Comment:
I think the `main_program` should have `feed_op`s and `fetch_op`s. However, how the `feed_op`s and `fetch_op`s come depends on the storing format. It may be inserted in the c++ code or may be initialized from protobuf message file.